### PR TITLE
Rerun a parameterized test of an abstract class within a test hierarchy

### LINKED ISSUE #21: cannot rerun failed tests of super class with parameterized test method
Within a test hierarchy a parameterized test method of a super class cannot be rerun in case of failure. 

```
 @spock.lang.Unroll
  abstract class AbstractTest extends spock.lang.Specification {
          def "unrolled [#param] parent"() {
               expect:
               ${flakyAssert()}

               where:
               param << ['foo']
                }
            }
            
class InheritedTest extends AbstractTest {

           def "inherited"() {
               expect:
               true
            }
         }

```

results in 
```
* What went wrong:
Execution failed for task ':test'.
> org.gradle.test-retry was unable to retry the following test methods, which is unexpected. Please file a bug report at https://github.com/gradle/test-retry-gradle-plugin/issues
     acme.InheritedTest#unrolled [foo] parent
```

### PR DESCRIPTION
**Pull request by jonnybbb**
Fixes #21

### CODE REVIEW

File: plugin/src/main/java/org/gradle/testretry/internal/RetryTestFrameworkGenerator.java [hunk 1]
```diff
@@ -151,7 +151,7 @@ private static boolean isSpockStepwiseTest(JvmTestExecutionSpec spec, TestName f
                     .map(testClass -> {
                         try (FileInputStream testClassIs = new FileInputStream(testClass)) {
                             ClassReader classReader = new ClassReader(testClassIs);
-                            SpockParameterClassVisitor visitor = new SpockParameterClassVisitor(failedTest.getName());
+                            SpockParameterClassVisitor visitor = new SpockParameterClassVisitor(failedTest.getName(),spec);
```
Comment (jonnybbb) [inline plugin/src/main/java/org/gradle/testretry/internal/RetryTestFrameworkGenerator.java]: TestName extraction should go into yhe SpockParameterClassVisitor completly

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java [hunk 2]
```diff
@@ -80,6 +95,25 @@ private static String normalizeMethodName(String methodPattern) {
         return methodPattern.replaceAll(SPOCK_PARAM_PATTERN, PARAM_PLACEHOLDER);
     }
 
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+        super.visit(version, access, name, signature, superName, interfaces);
+        if (superName != null && !superName.equals("java/lang/Object")) {
+            final FileCollection collection = spec.getTestClassesDirs().filter(d -> Files.exists(parentClass(d, superName)));
```
Comment (jonnybbb) [inline plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]: can we even make better use of the gradle JvmTestExecutionSpec (eg `JvmTestExecutionSpec.getPreviousFailedTestClasses()`). The `org.gradle.api.internal.tasks.testing.junit.JUnitDetector` knows already about the superClasses

**Comment by ldaley** [comment]
Solution needs to be more general. I'm going to close this one out. We can still use the code as guidance in the more general solution.

**Comment by jkschneider** [comment]
> Solution needs to be more general.

In fact this isn't necessary. Spock is the only framework for which we need to resort to ASM to look at superclass hierarchy. Added tests for all other frameworks in 92ae3adfe5da93a63146dc643055e50931dec5a3.

@jonnybbb had identified in one of the commits the desire to generalize how to locate the bytecodes for a Spock test's base class (from assuming it is in the same test source set to retrieving it from other places in the classpath). This was achieved in f5b5978180cb351e11d67df799dfa15743c37e97.

**Pull request closed without merging**